packages/blocks/src/__internal__/utils/selection.ts

```diff
@@ -532,36 +402,93 @@ export function handleNativeRangeDragMove(
+
   // Range from current mouse position
-  const currentRange = caretRangeFromPoint(e.raw.clientX, e.raw.clientY);
+  let currentRange = caretRangeFromPoint(e.raw.clientX, e.raw.clientY);
   if (!currentRange) return;
 
   assertExists(startRange);
   const { startContainer, startOffset, endContainer, endOffset } = startRange;
-  const container = currentRange.commonAncestorContainer as HTMLElement;
+  const _startContainer = (
+    startContainer.nodeType === Node.TEXT_NODE
+      ? startContainer.parentElement
+      : startContainer
+  ) as HTMLElement;
+  const startFrame = _startContainer.closest('affine-frame');
+  if (!startFrame) return;
+
+  const { clientX: x, clientY: y } = e.raw;
+
+  let currentFrame: FrameBlockComponent | null | undefined = null;
+  let shouldUpdateCurrentRange = false;
+
+  if (isEdgelessMode) {
+    currentFrame = startFrame;
+    shouldUpdateCurrentRange = true;
+  } else {
+    const el = document.elementFromPoint(x, y);
+    if (el?.classList.contains('quill-container')) {
+      return;
+    }
+    currentFrame = el?.closest('affine-frame');
+    const currentEditor = el?.closest('.ql-editor');
+    // if we are not pointing at an editor, we should update the current range
+    // if we are not even pointing at a frame, we should find one and update the current range
+    shouldUpdateCurrentRange = !currentFrame || !currentEditor;
+    currentFrame ??= getClosestFrame(y);
+  }
+  if (!currentFrame) return;
+
+  if (shouldUpdateCurrentRange) {
+    const closestEditor = getClosestEditor(y, currentFrame);
+    if (!closestEditor) return;
+
+    const newPoint = normalizePointIntoContainer({ x, y }, closestEditor);
+    currentRange = caretRangeFromPoint(newPoint.x, newPoint.y);
+    if (!currentRange) return;
+    if (!currentFrame.contains(currentRange.endContainer)) return;
+  }
+
   // Forward: ↓ →, Backward: ← ↑
   const isBackward = currentRange.comparePoint(endContainer, endOffset) === 1;
+  handleInFrameDragMove(
+    startContainer,
+    startOffset,
+    endContainer,
+    endOffset,
+    currentRange,
+    isBackward
+  );
+}
 
-  // Handle native range state on cross-block dragging,
-  // see https://github.com/toeverything/blocksuite/pull/845
-  if (container.nodeType === Node.TEXT_NODE) {
-    handleInFrameDragMove(
-      startContainer,
-      startOffset,
-      endContainer,
-      endOffset,
-      currentRange,
-      isBackward
-    );
+/**
+ * This function is used to normalize the point into the reasonable range of the container.
+ *
+ * It will set the point to the top-left or bottom-right corner
+ * when the point is out of the horizontal range of container.
+ */
+function normalizePointIntoContainer(point: IPoint, container: Element) {
+  const { top, left, right, bottom } = container.getBoundingClientRect();
+  const newPoint = { ...point };
+  const { x, y } = point;
+
+  // need this offset to avoid the point is out of the container
+  if (y < top) {
+    newPoint.y = top + 4;
```

Comment:
why 4?

Reply:
clicking on the edge of block will cause unstable behavior 😢